remote/src/main/java/com/spotify/metrics/remote/OkRemote.java

```diff
@@ -89,4 +91,17 @@ public void onResponse(Call call, Response response) throws IOException {
         });
         return result;
     }
+
+    @Override
+    public void waitForAllCalls() {
+        Dispatcher dispatcher = client.dispatcher();
```

Comment:
`Dispatcher` is not thread-safe. There's no guarantee that updated `queuedCallsCount`, `runningCallsCount` is even visible in the thread calling this method. (they are currently implemented as `ArrayDeque#size()` calls as per [Dispatcher.java](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/Dispatcher.java)).

The following is a suggestion to a working (maybe) working approach:

* Put the post in a synchronized section and maintain a count in it.
* When shutdown is reached, set a flag that disallows future requests.
* Take the count in a synchronized section and build a CountdownLatch.
* Block on `CountdownLatch#await()`.
* Put the Callback methods in a synchronized section, check for the presence of the latch and call `CountdownLatch#countDown()` if present.

A potentially more efficient approach would be to bound the number of requests using a `new Semaphore(Integer.MAX_VALUE)` and at shutdown block with `semaphore.acquire(Integer.MAX_VALUE, true)` (fairness is important). Each call to `post` would be guarded with a `semaphore#tryAcquire(1)`.